### PR TITLE
Add files via upload

### PR DESCRIPTION
**Pull request by gagc056**

### CODE REVIEW

File: enumerable.rb [hunk 1]
```diff
@@ -0,0 +1,90 @@
+# frozen_string_literal: true
+
+module Enumerable
+  def my_each
+    i = 0
+    while i < check_self.length
+      yield check_self[i]
```
Comment (samgaco) [inline enumerable.rb]: - [ ] my_each returns an Enumerator if no block is given

expected:

```
irb(main):167:0> [1,2].each
=> #<Enumerator: [1, 2]:each>
```

File: enumerable.rb [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    while i < check_self.length
+      yield check_self[i]
+      i += 1
+    end
+  end
+
+  def my_each_with_index
+    return to_enum :my_each_with_index unless block_given?
+
+    i = 0
```
Comment (samgaco) [inline enumerable.rb]: - [x] my_each_with_index calls the given block once for each element in self

```
irb(main):177:0> [1,2,"b","c"].each_with_index {|n,e| puts n,e}
1
0
2
1
b
2
c
3
=> [1, 2, "b", "c"]
```

File: enumerable.rb [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+    Enumerator.new do |y|
+      my_each { |item| y << item if yield item }
+    end.to_a
+  end
+
+  def my_all?(initial = nil)
+    result = true
+    if block_given?
+      my_each { |element| result &= (yield element) }
```
Comment (samgaco) [inline enumerable.rb]: Please test your methods against the following, which you can find [here](https://ruby-doc.org/core-2.6.5/Enumerable.html)

```
%w[ant bear cat].all? { |word| word.length >= 3 } #=> true
%w[ant bear cat].all? { |word| word.length >= 4 } #=> false
%w[ant bear cat].all?(/t/)                        #=> false
[1, 2i, 3.14].all?(Numeric)                       #=> true
[nil, true, 99].all?                              #=> false
[].all?                                           #=> true
``` 

this is what is exactly failing with your method

- [x] my_all? returns true if the block never returns false or nil
- [x] my_all? when no block or argument is given returns true when none of the collection members are false or nil
- [x] my_all? when a class is passed as an argument returns true if all of the collection is a member of such class
- [x] my_all? when a Regex is passed as an argument returns true if all of the collection matches the Regex
- [x] my_all? when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern

File: enumerable.rb [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+    else
+      my_each { |element| result &= element }
+    end
+    result
+  end
+
+  def my_any?
+    my_each { |item| return true if block_given? ? yield(item) : item }
+    false
+  end
```
Comment (samgaco) [inline enumerable.rb]: test my_any? against:

 ```
%w[ant bear cat].any? { |word| word.length >= 3 } #=> true
%w[ant bear cat].any? { |word| word.length >= 4 } #=> true
%w[ant bear cat].any?(/d/)                        #=> false
[nil, true, 99].any?(Integer)                     #=> true
[nil, true, 99].any?                              #=> true
[].any?                                           #=> false
```

- [ ] my_any? returns true if the block ever returns a value other than false or nil
- [ ] my_any? when no block or argument is given returns true if at least one of the collection is not false or nil
- [ ] my_any? when a class is passed as an argument returns true if at least one of the collection is a member of such class
- [ ] my_any? when a Regex is passed as an argument returns false if none of the collection matches the Regex

File: enumerable.rb [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+
+  def my_none?
+    my_each do |item|
+      return false if block_given? ? yield(item) : item
+    end
```
Comment (samgaco) [inline enumerable.rb]:
```
%w{ant bear cat}.none? { |word| word.length == 5 } #=> true
%w{ant bear cat}.none? { |word| word.length >= 4 } #=> false
%w{ant bear cat}.none?(/d/)                        #=> true
[1, 3.14, 42].none?(Float)                         #=> false
[].none?                                           #=> true
[nil].none?                                        #=> true
[nil, false].none?                                 #=> true
[nil, false, true].none?                           #=> false
```

- [x] my_none? returns true if the block never returns true for all elements
- [x] my_none? when no block or argument is given returns true only if none of the collection members is true
- [x] my_none? when a class is passed as an argument returns true if none of the collection is a member of such class
- [x] my_none? when a Regex is passed as an argument returns true only if none of the collection matches the Regex
- [x] my_none? when a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern

File: enumerable.rb [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+    true
+  end
+
+  def my_count(item = nil)
+    count = 0
+    if block_given?
+      my_each { |i| count += 1 if yield(i) == true }
```
Comment (samgaco) [inline enumerable.rb]:
```
ary = [1, 2, 4, 2]
ary.count               #=> 4
ary.count(2)            #=> 2
ary.count{ |x| x%2==0 } #=> 3
```

- [x] my_count returns the number of items in enum through enumeration
- [x] my_count counts the number of items in enum that are equal to item if an argument is given
- [x] my_count counts the number of elements yielding a true value if a block is given

File: enumerable.rb [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+    else
+      my_each { |i| count += 1 if i == item }
+    end
+    count
+  end
+
+  def my_map
+    return to_enum :my_map unless block_given?
+
+    arr = []
```
Comment (samgaco) [inline enumerable.rb]: - [x] my_map returns a new array with the results of running block once for every element in enum.

this should work:

 ```
irb(main):180:0> [1,2,3].map {|n| n*2}
=> [2, 4, 6]

```

File: enumerable.rb [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+    my_each { |element| arr << (yield element) }
+    arr
+  end
+
+  def my_inject(acc = nil, cur = nil)
+    tmp = is_a?(Range) ? to_a : self
+    a = acc.nil? || acc.is_a?(Symbol) ? tmp[0] : acc
+    if block_given?
```
Comment (samgaco) [inline enumerable.rb]: - [x] my_inject combines all elements of enum by applying a binary operation, specified by a block
- [x] my_inject when a symbol is specified combines each element of the collection by applying the symbol as a named method

```
# Sum some numbers
(5..10).reduce(:+)                             #=> 45
# Same using a block and inject
(5..10).inject { |sum, n| sum + n }            #=> 45
# Multiply some numbers
(5..10).reduce(1, :*)                          #=> 151200
# Same using a block
(5..10).inject(1) { |product, n| product * n } #=> 151200
# find the longest word
longest = %w{ cat sheep bear }.inject do |memo, word|
   memo.length > word.length ? memo : word
end
longest                                        #=> "sheep"
```

**Review by samgaco — CHANGES_REQUESTED** [review]
Hi @lhscar56 :high_brightness: 

I left you some comments, please test your methods against the cases I attached and try to make changes accordingly so your methods give the expected results. 

Also, I would require from you to improve the readme file.

This project is a bit challenging, so good spirit :muscle: 

----------------------

* ###### STATUS OF THE PROJECT: REQUESTED CHANGES :stop_sign:

**Review by whiz25 — APPROVED** [review]
## Awesome :clap:
You have made all the requested changes and your methods work as required. It's a :+1: from me.

### Martin Afani

**Pull request merged**